model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java

```diff
@@ -388,7 +388,7 @@ public void execute() {
             
             Predicate<? super V> entityFilter = mmcb.getEntityFilter();
             Predicate<? super K> keyFilter = ((MapModelCriteriaBuilder) mcb).getKeyFilter();
-            return v -> v != null && ! (keyFilter.test(v.getId()) && entityFilter.test(v));
+            return v -> v == null || ! (keyFilter.test(v.getId()) && entityFilter.test(v));
```

Comment:
This method returns a filter for objects that are not deleted by this `BulkDeleteOperation`. This is used to filter out pending tasks in the transaction that are not affected by this `BulkDeleteOperation`, and remove those that are (optimization on deletes).
`v` in this case is the entity after performing a pending operation, and in case of _delete_ operation this is `null`.
Hence bulk deletes were "optimized", i.e. replaced, by a subsequent `BulkDeleteOperation`, effectively ignoring the fact that the deleted objects could have been disjoint.
This is softening the condition: if the object has been removed by a previous operation, i.e. is `null`, do not remove it. 
This fixes functional bug at a little cost of performance.

Reply:
I see. I think I got it. Thanks!